homeassistant/components/rfxtrx/__init__.py

```diff
@@ -110,18 +109,24 @@ def handle_receive(event):
         # Log RFXCOM event
         if not event.device.id_string:
             return
-        _LOGGER.debug(
-            "Receive RFXCOM event from "
-            "(Device_id: %s Class: %s Sub: %s, Pkt_id: %s)",
-            slugify(event.device.id_string.lower()),
-            event.device.__class__.__name__,
-            event.device.subtype,
-            "".join(f"{x:02x}" for x in event.data),
-        )
+
+        event_data = {
+            "packet_type": event.device.packettype,
+            "sub_type": event.device.subtype,
+            "type_string": event.device.type_string,
+            "id_string": event.device.id_string,
+            "data": "".join(f"{x:02x}" for x in event.data),
+            "values": getattr(event, "values", None),
+        }
+
+        _LOGGER.debug("Receive RFXCOM event: %s", event_data)
 
         # Callback to HA registered components.
         hass.helpers.dispatcher.dispatcher_send(SIGNAL_EVENT, event)
 
+        # Signal event to any other listeners
+        hass.bus.fire(EVENT_RFXTRX_EVENT, event_data)
```

Comment:
We shouldn't just fire all events for the sake of firing events. All events that are handled by entities, we should probably just let it be represented as state changes. We should only try to fire events that are "event"-like, like pressing buttons, motion detectors etc.

Reply:
It's somewhat hard to know. Also there are stuff that get exposed as sensors, which sort of are a button presses (Chimes). They show up as a sensor indicating which chime sound.

Also these things are assumed state stuff often, since we are mostly listening to the remote commands and or generating it. No feedback. So if entity is in the wrong state, there will be no change event on the entity for a rfxtrx event.

Reply:
Doorbell presses / chimes should really be an event. You can have a separate sensor that just keeps track of the configured chime if that is important. 

If state out of sync is an issue, consider using the `RestoreState` functionality.

Reply:
My problem I don't really know which event are what. I have one doorbell and one set of switches. I won't be able to make much of an educated guess as to which should be events. If @danielhiversen is still around maybe he can chime in on that. Doorbells for example have no high-level abstraction. They come through just as a sensor with a value for chime sound.

Without a deep understanding of the underlying lib. I can't translate/filter these events. 

Even if we do parse some stuff into entities , the event contains other values that (maybe) we don't take into account.

Reply:
The 433MHz protocol is used by many types of sensors and it is not standardized as z-wave, and the pyRFXtrx will never be able to parse all types of sensors/devices packages. 

Another option is to set the event_data as an attribute for the device.

Reply:
At this stage there are no home assistant devices linked. Also i don't think we support configuration per device yet do we? So we can't really move the setting there. Also, event_data on the device won't work since we can't trigger on that.

Reply:
So should we then treat this more like MQTT? 

MQTT has a subscribe function and their own automation trigger. 

Just trying to avoid having all the rfxtrx events be on our event bus especially if they might not be used.

Reply:
These devices don't seem especially chatty. They are usually on battery power. I get one message every other second in total with 3 weatherstations. My fire alarms are silent by default. Switches are only on button press. Sure there could be devices out there with a lot of comm. But...

Reply:
> 
> 
> It's somewhat hard to know. Also there are stuff that get exposed as sensors, which sort of are a button presses (Chimes). They show up as a sensor indicating which chime sound.
> 
> Also these things are assumed state stuff often, since we are mostly listening to the remote commands and or generating it. No feedback. So if entity is in the wrong state, there will be no change event on the entity for a rfxtrx event.

If missing an event because entity is in wrong state is your main issue here, the "force_update" property would help. This is anyway I property that is useful when e.g. sending data to a database like influxdb and each received rfx event should lead to a database entry.

Reply:
Not really database that is the case here. Its triggering actions on events.

Reply:
I know, however the state update event is normally blocked when the state does not change. For your case that means if the entity is in the wrong state and an event from rfxtrx does not lead to a state change you won't get a state update event. Setting the force_update to true always raises the state changed event even if the state doesn't change.

My database case also always needs state update events, otherwise no data is written in the database. I'm just adding an argument here where I think both use cases can be solved with the force_update property set optionally to true.

Reply:
So basically, if you want to set an action on an event, you can just use the state changed event. When the force_update is true, writing the state from an rfxtrx event will always fire the state changed event.

Reply:
And it also brings us back to the very first comment of @balloob 
> we should probably just let it be represented as state changes
As far as I can see, the only issue you have with that is that state changes are not always fired because of above mentioned reason, which again should be fixed by having the option per entity to set a force_update to true.

I agree with previous comments that firing all rfxtrx events on a bus could lead to a lot of events. Do also note that the device does not only receive data from your devices, it receives data from any device in range for which protocols are enabled. Having the force_update per entity, you can select per entity where it makes sense to forward every rfxtrx event as state changed event on the event bus

Reply:
once we move to config flow, there will be no per-entity config. So i can't really do it that way.

Reply:
I guess it can't hurt to enable it globally for the integration as an option.

In a config flow the question would be how to handle discovery and perhaps delete discovered entities. If for every discovered device and every attribute an entity is made and an event is fired that will fill the bus rapidly. Some discovered devices may not be yours anyway. As long as you end up with only the entities you really need I don't see how globally setting it would be an issue.

Question is then how is discovery dealt with and how to get rid of unwanted entities that were discovered but either not needed or not your devices.

Reply:
Also note, we are really representing two different things described here. One thing being the actuator and one being the sensor. You present the button press on the sensor as an event. You represent the actuator (doorbell, light switch) as a state.

You very much can have use for either of these things separate. The button can control scene:s in home assistant (without it ever being linked to any physical actuator) while the wallsocket switch can be controlled from home assistant, without there ever being a physical button.

Reply:
But still, the button press can be represented as 'state'? You could state it as triggered and after some time with no updates set it back to idle or something like that. When force_update is true, the some time is only for graphical representation and even when you click it twice very fast you would get all events. You can set your automation to look for new state 'triggered'?

I didn't do a really deep dive here yet, honestly speaking I see you run into similar issues as I do and see some good changes here already!

Do note that when not setting polling to false, a force_update will lead to wrong events because the state will also be updated from the polling loop.

Reply:
the binary_sensor actually have that delay_off feature. Where it can revert back. (doorbell chime:s are oddly represented as SensorEvent's, so they don't get allowed into the binary_sensor integration at the moment). Also the delay_off feature was requested to be removed. I was also thinking along the line of auto-reverting momentary sensors.

Once config_flow and device registry is in. All devices, could get auto created binary_sensors (disabled by default) that autorevert directly. Ie they write state twice. But we are far from getting there with my config_flow pull :)

Reply:
Button presses can be made events while sensors are states. Both can then be responded to by automations. Since states are based on events, using restore state to restore the latest state seems reasonable.

Reply:
To be clear, the auto revert feature indicates either that the device is stateless and then should not be represented by an entity in our architecture design but custom events or that the wrong entity platform representation is used since the entity state isn't really the device state.

Push buttons, door chimes etc should not be represented by binary sensors in home assistant unless they report two discrete states.

Reply:
Generally having buttons as events and sensors as state is a good solution. I think what @elupus already mentions the main issue here is how to know whether it's a button or a sensor, because from the data it cannot always be observed. Perhaps the solution lies within the used protocol. For example, AC/ARC seems to be only used by "event" devices like buttons but also smoke detectors. Sensors like temperature sensors which report useful data have their own protocol, like Oregon, AE and many more.

The RFXtrx user guide contains a list of supported devices and which protocol they use. It may be needed to go though them and categorize it (and check if my statement on protocol being either event or state is true)

Reply:
By the way, you're taking on quite some work here and with the whole config_flow. If you need some assistance, I'm happy to step in :)

Reply:
I'm sadly close to giving up on this. I won't be able to deduce what things are. I have way too few devices and I have no interest breaking people's setups.

Dropping delay_off for example would break all motion sensors. They only report motion. Never no motion. They report this periodically. And I don't have any to test.

Dropping events for switches, breaks people using the remotes for such switches for other actions in HA ( they are really two things )

Reply:
Here's my two cents. Currently the way it's working is to me strange, because data is received from the rfxtrx in an event manner and the state in home assistant is updated using polling. To get events to the user there is a "fire_event" but actually if you have a real sensor and link an automation to that, the state is not updated yet because the rfxtrx integration does not bother about updating states. That makes the fire_event for automations that are based on sensors useless.

In my opinion, a first useful step is to update the state from the rfxtrx event (which you did), the should_poll should be a property then returning False, because data is updated with push and not poll. That makes the whole state updating of the sensor way more logical.

Next step would be how to deal with the fire_event. Although maybe not the nicest solution, as long as buttons etc at least result in one entity with a state, I think the state update events could be a good substitute for this event, given that the force_update property is set. How to do that is another matter, because yaml changes are not allowed for platform integrations. I see two ways here:
- Allow the yaml change to be able to close this PR, give users the chance to update their automations (and get user feedback) and then do the other PR to make a configuration flow.
- Finish them both in parallel, where this would give only a partial solution and adding a config_flow where somehow the force_update property can be set gives a full solution and both should be released at the same time.

This would mean not using events for buttons but state updates (with respective events), may not be ideal but it's not really different from what we do now, except that the fire_event is replaced by a state update event. Future work may be to find out how to distinguish between devices and make events for buttons

Feel free to comment on it.

Reply:
I see I am not fully up to speed...
The first "step" you mention you already tackled with PR #37271
Do note that the state is still also polled, because as far as I can see, you didn't add a property "should_poll" which returns false.

Reply:
Doesn't matter what that return. It's not polled It's event driven. That property just influences if the update method is called periodically. It will schedule an update on event.

Reply:
A bit unrelated to this PR, but why let hass periodically call the update method if you also schedule it on event? The periodic call will never update the state, so it can't hurt but it's also not useful.

Reply:
I really thought it was disabled

Reply:
Check out entity_platform.py:

`async with self._process_updates:
            tasks = []
            for entity in self.entities.values():
                if not entity.should_poll:
                    continue
                tasks.append(entity.async_update_ha_state(True))  # type: ignore

            if tasks:
                await asyncio.gather(*tasks)`

By default should_poll is True (see entity.py). As I said, makes no sense here to have it to True.

Reply:
It's fine to have a motion sensor exist with a delay off. But in that case we shouldn't fire events. I think the main thing here is that messages are either consumed by an entity or fired on the event bus.

Reply:
The former, delay off, is interfering with our entity model. The latter, custom events, are not interfering as the consumption of those is optional.

Reply:
I think that event + delay off is ok to store in a binary sensor, since together they provide all the information necessary to be a binary sensor.

Reply:
Then we're not following our own guidelines that stateless devices should not be entities.

We're making it harder for ourselves when making exceptions to our guidelines. I think we should avoid that at almost all cost. A breaking change for an integration is ok if it means bringing the integration in line with the rest of the ecosystem and still providing useful features to the users.

Reply:
Well depends on what you meen. They are not really stateless. Just that the state is hidden in the device, and only communicated one way.

Reply:
Another standard that we have is that we should model integrations after the information provided by the device or service and not add custom logic on top of that even to mimic some behavior of the device.

Reply:
i don't follow that really. So any device which only signals data in active state and stays quiet in non active state can not be represented by an entity in home assistant? If if it's fully by design to repeat "i'm on" every X second, and not send anything when off.

Reply:
I think that it depends on who supplies the "turn off" value. If the user has to set it in YAML/UI, it means they are converting an event into a binary sensor. That could also be provided as a generic solution for any event.

Reply:
How generic do you have in mind?

Reply:
Home Assistant event generic. Configure binary sensors with same matching as automation event trigger + a time off period.

Reply:
If I get back to original topic of this pull ;). I think I have a solution now.

Once I moved to integration level config, it occured to me. I'm going to change the config to be per device rather than platform. So you setup the config option fire event, per packet which represent the device and can generate many entities. That way we can fire events for specific devices only and config is simplified.